### ISSUE TITLE
Change in NodePort for LoadBalancer services causes traffic drop in presence of cloud provider errors

### ISSUE THREAD

**Issue by ankurchoubeymsft** [kubernetes/kubernetes#103622]
#### What happened:
If there is a change in the NodePort for a service with type=LoadBalancer, and the EnsureLoadBalancer operation failed to update cloudprovider's load balancer with the new NodePort, the iptables rules for old NodePort are still removed causing a 100% drop in availability/connectivity of the external facing IP address leading to a service outage. 

The kubeproxy updated the iptables rules which only contained the rules for new NodePort (chain KUBE-NODEPORTS under iptable nat). But the cloud provider load balancer is still probing the old NodePort as the EnsureLoadBalancer operation did not complete. Since iptables rules do not contain any information about older NodePort, the probing traffic from cloud provider was dropped and all instances were marked unhealthy. 

The NodePort is randomly chosen by K8s if the Service spec does not contain one (most common case). For any update operation with the same Service spec without any changes (i.e. reconcile), the NodePort still changes. The EnsureLoadBalancer operation triggered due to this change can fail due to several reasons such as rate limiting by cloud provider, authentication failures or other cloud-provider side server errors. The underlying reason for the EnsureLoadBalancer failure can take several minutes to hours for resolution but until then the traffic is dropped 100%.

#### What you expected to happen:
Even if cloudprovider update fails (control plane operation), the traffic flow should continue to work (data plane). 

Until the cloud provider load balancer can be successfully updated, the old NodePort should continue to work so that the traffic can still flow. Kubeproxy should keep the iptables rules in chain KUBE-NODEPORTS for both old and new NodePort and only when the load balancer operation completes, delete the rules for old NodePort. 

#### How to reproduce it (as minimally and precisely as possible):
1. Have a working k8s service with type=LoadBalancer. The service spec should not contain NodePort so that K8s can choose one randomly (most common case). Ensure traffic flows to this public IP address. kubectl describe the service to check the NodePort assigned. On any node, check iptables rules (```sudo iptables -t nat -L KUBE-NODEPORTS```) corresponding to this service/port number and note the NodePort used.
2. Mimic update load balancer operation failure as suitable for your test environment. Some examples may include reducing rate limiting threshold to 0 causing rate limiting errors, or cut access to the cloud provider by temporarily removing authentication policies.
3. Trigger an update to NodePort for this service. The service spec should also not contain the NodePort so that K8s can choose one randomly. 
4. Check the service details to see the change in NodePort. Events for the service will show errors related to EnsureLoadBalancer operation. 
Check the cloud provider's portal to see the probe port which should still point to the old NodePort.
The traffic flow will stop at the same time (once kubeproxy gets the update). Check iptables for KUBE-NODEPORTS chain which contains only the new NodePort (```sudo iptables -t nat -L KUBE-NODEPORTS```).
5. Now, unblock the access to perform load balancer operations. The cloud provider's load balancer probe port will be updated and traffic starts flowing again.

#### Environment:
- Kubernetes version (use `kubectl version`): 1.18
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): Ubuntu 18.04
- Kernel (e.g. `uname -a`): 5.4.0-1051-azure
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others: Istio (that creates/reconciles the service)

/sig network

**Comment by aojea** [comment]
/sig cloud-provider
This look more like a cloud-provider problem than a sig-network problem, if I understand if correctly, if you have already updated the service there is no way you can keep the old information working, because is not part of the service anymore ...

**Comment by feiskyer** [comment]
this is an expected behavior and cloud provider would retry on the errors.

Because of throttling and potentially failures from LB API, you could use a fixed NodePort to reduce number of updates (e.g. enabling Local traffic policy and use patch method for later updates).
